File: fragments/sqlite_db/sqlite_crud.test.ts
import { addItem, deleteItem, getAllItems, getItemByTicker, getItemsAfterLaunchYear, updateItem } from "./sqlite_crud";
import assert from "node:assert/strict";
import test, { describe } from "node:test";
import { randomUUID } from "node:crypto";

describe("sqlite crud", () => {
  describe("getting by ticker", () => {
    test("retrieves a known ticker", () => {
      const result = getItemByTicker("BTC");
      assert.strictEqual(result!.ticker, "BTC");
      assert.strictEqual(result!.name, "Bitcoin");
      assert.strictEqual(result!.launched, 2009);
    });

    test("handles an unknown ticker", () => {
      const result = getItemByTicker(randomUUID().slice(0, 6).toUpperCase());
      assert.strictEqual(result, undefined);
    });
  });

  describe("get items after launch year", () => {
    test("matching items after launch year", () => {
      const result = getItemsAfterLaunchYear(2000);
      assert.ok(result.length >= 3);
    });

    test("no matching items after launch year", () => {
      const result = getItemsAfterLaunchYear(2050);
      assert.strictEqual(result.length, 0);
    });
  });

  test("get all items ordered by launch year", () => {
    const result = getAllItems();
    assert.ok(result.length >= 3);
  });

  describe("adding items", () => {
    test("adds an item to the database table", () => {
      const ticker = randomUUID().slice(0, 6).toUpperCase();
      const result = addItem({
        ticker,
        name: "Test Coin",
        launched: 2023,
      });
      assert.strictEqual(result, "ok");

      const item = getItemByTicker(ticker);
      assert.strictEqual(item!.ticker, ticker);
    });

    test("handles duplicates", () => {
      const ticker = randomUUID().slice(0, 6).toUpperCase();
      addItem({
        ticker,
        name: "Test Coin",
        launched: 2023,
      });

      const result = addItem({
        ticker,
        name: "Test Coin Duplicate",
        launched: 2024,
      });
      assert.strictEqual(result, "ok");

      const item = getItemByTicker(ticker);
      assert.strictEqual(item!.name, "Test Coin");
      assert.strictEqual(item!.launched, 2023);
    });
  });

  describe("updating items", () => {
    test("updates existing item", () => {
      const ticker = randomUUID().slice(0, 6).toUpperCase();
      addItem({
        ticker,
        name: "Original Name",
        launched: 2023,
      });

      const result = updateItem({
        ticker,
        name: "Updated Name",
        launched: 2024,
      });
      assert.strictEqual(result!.ticker, ticker);
      assert.strictEqual(result!.name, "Updated Name");
      assert.strictEqual(result!.launched, 2024);
    });

    test("updates nonexistent item", () => {
      const ticker = randomUUID().slice(0, 6).toUpperCase();
      const result = updateItem({
        ticker,
        name: "Nonexistent",
        launched: 2012,
      });
      assert.strictEqual(result, undefined);
    });
  });

  describe("deleting items", () => {
    test("deletes existing item", () => {
      const ticker = randomUUID().slice(0, 6).toUpperCase();
      addItem({
        ticker,
        name: "To Be Deleted",
        launched: 2024,
      });

      const result = deleteItem(ticker);
      assert.strictEqual(result!.ticker, ticker);

      const check = getItemByTicker(ticker);
      assert.strictEqual(check, undefined);
    });

    test("deletes nonexistent item", () => {
      const ticker = randomUUID().slice(0, 6).toUpperCase();
      const result = deleteItem(ticker);
      assert.strictEqual(result, undefined);
    });
  });
});

File: fragments/sqlite_db/sqlite_crud_test.go
package sqlite_crud

import (
	"testing"
)

func TestGetItemByTicker(t *testing.T) {
	result, _ := GetItemByTicker("BTC")

	if result == nil {
		t.Error("Expected coin 'BTC' to exist")
	}
}

func TestGetItemByTicker_NotFound(t *testing.T) {
	result, _ := GetItemByTicker("XRP")

	if result != nil {
		t.Error("Expected query for XRP to be 'nil'")
	}
}

func TestGetItemsAfterLaunchYear(t *testing.T) {
	result, _ := GetItemsAfterLaunchYear(2000)

	if len(result) != 3 {
		t.Errorf("Expected 3 results, but got %d", len(result))
	}
}

func TestGetItemsAfterLaunchYear_NoResults(t *testing.T) {
	result, _ := GetItemsAfterLaunchYear(2020)

	if len(result) != 0 {
		t.Errorf("Expected 0 results, but got %d", len(result))
	}
}

func TestGetAllItems(t *testing.T) {
	result, _ := GetAllItems()

	if result[0].Ticker != "SOL" {
		t.Errorf("Expected first item to be SOL, but got %s", result[0].Ticker)
	}
	if result[1].Ticker != "ETH" {
		t.Errorf("Expected first item to be ETH, but got %s", result[1].Ticker)
	}
	if result[2].Ticker != "BTC" {
		t.Errorf("Expected first item to be BTC, but got %s", result[2].Ticker)
	}
}

func TestAddItem(t *testing.T) {
	coin := CryptoCoinWithoutId{Ticker: "PEPE", Name: "Pepe", Launched: 2023}
	_, newId, _ := AddItem(coin)

	if newId != 4 {
		t.Errorf("Expected 'newId' to be 4, but got %d instead", newId)
	}
}

func TestAddItemDuplicate(t *testing.T) {
	coin := CryptoCoinWithoutId{Ticker: "BTC", Name: "Bitcoin", Launched: 2009}
	_, newId, _ := AddItem(coin)

	if newId != 3 {
		t.Errorf("Expected 'newId' to be 3, but got %d instead", newId)
	}
}

func TestUpdateItem(t *testing.T) {
	coin := CryptoCoinWithoutId{Ticker: "BTC", Name: "Bitcoin", Launched: 2008}
	result, _ := UpdateItem(coin)

	if result == nil {
		t.Error("Expected update to return coin")
		t.FailNow()
	}

	if result.Ticker != "BTC" {
		t.Errorf("Expected ticker to be BTC, but got %s", result.Ticker)
	}
}

func TestUpdateItem_NotFound(t *testing.T) {
	coin := CryptoCoinWithoutId{Ticker: "XRP", Name: "Ripple", Launched: 2012}
	result, _ := UpdateItem(coin)

	if result != nil {
		t.FailNow()
	}
}

func TestDeleteItem(t *testing.T) {
	result, _ := DeleteItem("ETH")

	if result == nil {
		t.Error("Expected delete to return coin")
		t.FailNow()
	}

	if result.Ticker != "ETH" {
		t.Errorf("Expected ticker to be ETH, but got %s", result.Ticker)
	}
}

func TestDeleteItem_NotFound(t *testing.T) {
	result, err := DeleteItem("XRP")

	if result != nil {
		t.FailNow()
	}
	if err != nil {
		t.FailNow()
	}
}
